src/clip.sol

```diff
@@ -0,0 +1,472 @@
+
+            // DAI needed to buy a slice of this sale
+            owe = mul(slice, price);
+
+            // Don't collect more than tab of DAI
+            if (owe > tab) {
+                // Total debt will be paid
+                owe = tab;                  // owe' <= owe
+                // Adjust slice
+                slice = owe / price;        // slice' = owe' / price <= owe / price == slice <= lot
+            } else if (owe < tab && slice < lot) {
+                // If slice == lot => auction completed => dust doesn't matter
+                uint256 _chost = chost;
+                if (tab - owe < _chost) {    // safe as owe < tab
+                    // If tab <= chost, buyers have to take the entire lot.
+                    require(tab > _chost, "Clipper/no-partial-purchase");
+                    // Adjust amount to pay
+                    owe = tab - _chost;      // owe' <= owe
+                    // Adjust slice
+                    slice = owe / price;     // slice' = owe' / price < owe / price == slice < lot
+                }
+            }
+
+            // Calculate remaining tab after operation
+            tab = tab - owe;  // safe since owe <= tab
+            // Calculate remaining lot after operation
+            lot = lot - slice;
+
+            // Send collateral to who
+            vat.flux(ilk, address(this), who, slice);
+
+            // Do external call (if data is defined) but to be
+            // extremely careful we don't allow to do it to the two
+            // contracts which the Clipper needs to be authorized
+            DogLike dog_ = dog;
+            if (data.length > 0 && who != address(vat) && who != address(dog_)) {
+                ClipperCallee(who).clipperCall(msg.sender, owe, slice, data);
+            }
+
+            // Get DAI from caller
+            vat.move(msg.sender, vow, owe);
+
+            // Removes Dai out for liquidation from accumulator
+            dog_.digs(ilk, lot == 0 ? tab + owe : owe);
+        }
+
+        if (lot == 0) {
+            _remove(id);
+        } else if (tab == 0) {
+            vat.flux(ilk, address(this), usr, lot);
+            _remove(id);
+        } else {
+            sales[id].tab = tab;
+            sales[id].lot = lot;
+        }
+
+        emit Take(id, max, price, owe, tab, lot, usr);
+    }
+
+    function _remove(uint256 id) internal {
+        uint256 _move    = active[active.length - 1];
+        if (id != _move) {
+            uint256 _index   = sales[id].pos;
+            active[_index]   = _move;
+            sales[_move].pos = _index;
+        }
+        active.pop();
+        delete sales[id];
+    }
+
+    // The number of active auctions
+    function count() external view returns (uint256) {
+        return active.length;
+    }
+
+    // Return the entire array of active auctions
+    function list() external view returns (uint256[] memory) {
+        return active;
+    }
```

Comment:
`active` has a `public` visibility modifier. This seems redundant. Why is it provided?

Reply:
No, `active()` expects an index and returns the value of that index. `list()` returns the whole array.